## Amazon SageMaker HyperParameterTuning 

In this notebook we will demonstrate the following the automated hyperparameter tuning capability with Amazon SageMaker.

### Overview

1. Set up
2. Automatic model tuning (HPT) of XGBoost algorithm for weather prediction
3. Compare Bayesian (default) and Random tuning strtegies

### Step 1 - Set up

##### 1.1 Import libraries

In [1]:
import boto3
import pandas as pd

import sagemaker
from sagemaker import image_uris
from sagemaker.session import Session
from sagemaker.inputs import TrainingInput

from sagemaker.tuner import (
    IntegerParameter,
    CategoricalParameter,
    ContinuousParameter,
    HyperparameterTuner,
)

##### 1.2 Define variables

In [2]:
# set an output path where the trained model will be saved
s3_bucket = 'datascience-environment-notebookinstance--06dc7a0224df'
s3_prefix = 'prepared'
m_prefix = 'hpo'
output_path = 's3://{}/{}/{}/output'.format(s3_bucket, m_prefix, 'xgboost')
content_type = 'text/csv'

sagemaker_session = sagemaker.Session()
region = sagemaker_session.boto_region_name

### Step 2 -  Automatic model tuning of the XGBoost algorithm

#### 2.1 Define parameters necessary for the HPT job

In [3]:
#Objective metric to optimize
objective_metric_name = "validation:rmse"
#Objective type.  Here we want to minimize the objective metric
objective_type = 'Minimize'
#Define the hyperparameter ranges to explore
hyperparameter_ranges = {
    "eta": ContinuousParameter(0, 1),
    "min_child_weight": ContinuousParameter(1, 10),
    "alpha": ContinuousParameter(0, 2),
    "max_depth": IntegerParameter(1, 10),
}
#Total training jobs to be run as part of HPT
max_jobs=10
#Maximum number of training jobs to run in parallel
max_parallel_jobs=2

#### 2.2 Define the estimator and HPT job

In [4]:
# This line automatically looks for the XGBoost image URI and builds an XGBoost container.
# specify the repo_version depending on your preference.
xgboost_container = sagemaker.image_uris.retrieve("xgboost", region, "1.2-1")

In [5]:
# initialize hyperparameters
hyperparameters = {
        "max_depth":"5",
        "eta":"0.2",
        "gamma":"4",
        "min_child_weight":"6",
        "subsample":"0.7",
        "objective":"reg:squarederror",
        "num_round":"5"}

In [6]:
# define the data type and paths to the training and validation datasets
train_input = TrainingInput("s3://{}/{}/{}/".format(s3_bucket, s3_prefix, 'train'), content_type=content_type, distribution='ShardedByS3Key')
validation_input = TrainingInput("s3://{}/{}/{}/".format(s3_bucket, s3_prefix, 'validation'), content_type=content_type, distribution='ShardedByS3Key')

In [7]:
#Define the estimator
estimator_hpo = sagemaker.estimator.Estimator(image_uri=xgboost_container, 
                                          hyperparameters=hyperparameters, ##Is this needed?
                                          role=sagemaker.get_execution_role(),
                                          instance_count=1, 
                                          instance_type='ml.m5.12xlarge', 
                                          volume_size=200, # 5 GB 
                                          output_path=output_path)

In [8]:
#Define the Tuner object
tuner = HyperparameterTuner(
    estimator_hpo, 
    objective_metric_name, 
    hyperparameter_ranges, 
    max_jobs=10,  #Total training jobs to be run as part of HPT
    max_parallel_jobs=2, #Maximum number of training jobs to run in parallel
    objective_type = objective_type
)

#### 2.3 Run the tuner job

In [9]:
#Fit method
tuner.fit({'train': train_input, 'validation': validation_input})

........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

In [10]:
tuning_job_name = tuner.latest_tuning_job.name
tuning_job_name

'sagemaker-xgboost-210807-2351'

#### 2.4 Examine the HPT results

In [11]:
##Get results from HPT
tuner_results = sagemaker.HyperparameterTuningJobAnalytics(tuning_job_name)
##Get the results as a dataframe
full_df = tuner_results.dataframe()
##Display results sorted by objective metric values
if len(full_df) > 0:
    df = full_df[full_df["FinalObjectiveValue"] > -float("inf")]
    if len(df) > 0:
        df = df.sort_values("FinalObjectiveValue", ascending=True)
        print("Number of training jobs with valid objective: %d" % len(df))
        print({"lowest": min(df["FinalObjectiveValue"]), "highest": max(df["FinalObjectiveValue"])})
        pd.set_option("display.max_colwidth", -1)  # Don't truncate TrainingJobName
    else:
        print("No training jobs have reported valid results yet.")

df

Number of training jobs with valid objective: 10
{'lowest': 0.478520005941391, 'highest': 0.6521400213241577}


/home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages/ipykernel/__main__.py:12: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.


alpha       eta  max_depth  min_child_weight  \
0  0.791758  1.000000  5.0        9.303584           
3  0.807994  0.859559  4.0        8.958240           
8  0.055109  0.165204  8.0        8.342793           
7  0.037281  0.753968  8.0        3.775813           
9  1.586089  0.602610  3.0        1.451194           
2  0.673939  0.177297  4.0        2.935260           
6  0.019703  0.104085  1.0        8.741634           
1  0.270679  0.871643  8.0        5.661067           
5  1.707530  0.676392  10.0       1.465015           
4  1.484935  0.044220  10.0       9.393024           

                              TrainingJobName TrainingJobStatus  \
0  sagemaker-xgboost-210807-2351-010-225ea843  Completed          
3  sagemaker-xgboost-210807-2351-007-bfbae99c  Completed          
8  sagemaker-xgboost-210807-2351-002-c0683818  Completed          
7  sagemaker-xgboost-210807-2351-003-acd79818  Completed          
9  sagemaker-xgboost-210807-2351-001-5def3b51  Completed          
2  sagemaker-xgboost-210807-2351-008-3ffac53d  Completed          
6  sagemaker-xgboost-210807-2351-004-fca96b08  Completed          
1  sagemaker-xgboost-210807-2351-009-47078b98  Completed          
5  sagemaker-xgboost-210807-2351-005-13210913  Completed          
4  sagemaker-xgboost-210807-2351-006-9b77dd2f  Completed          

   FinalObjectiveValue         TrainingStartTime           TrainingEndTime  \
0  0.47852             2021-08-08 01:31:16+00:00 2021-08-08 01:50:58+00:00   
3  0.50249             2021-08-08 01:10:16+00:00 2021-08-08 01:28:52+00:00   
8  0.50458             2021-08-07 23:54:03+00:00 2021-08-08 00:15:43+00:00   
7  0.51216             2021-08-08 00:15:14+00:00 2021-08-08 00:38:23+00:00   
9  0.53063             2021-08-07 23:53:45+00:00 2021-08-08 00:11:59+00:00   
2  0.55432             2021-08-08 01:10:13+00:00 2021-08-08 01:28:23+00:00   
6  0.60112             2021-08-08 00:18:34+00:00 2021-08-08 00:34:09+00:00   
1  0.64024             2021-08-08 01:31:08+00:00 2021-08-08 01:53:14+00:00   
5  0.64992             2021-08-08 00:39:40+00:00 2021-08-08 01:03:27+00:00   
4  0.65214             2021-08-08 00:40:54+00:00 2021-08-08 01:04:42+00:00   

   TrainingElapsedTimeSeconds  
0  1182.0                      
3  1116.0                      
8  1300.0                      
7  1389.0                      
9  1094.0                      
2  1090.0                      
6  935.0                       
1  1326.0                      
5  1427.0                      
4  1428.0

The tuning job took roughly 2 hours. Best validation rmse observed is 0.42 with max_depth = 6, alpha = 0.82, eta = 0.46, min_child_weight = 3.41.  

What are the best values you observed?

### Step 3 Compare Bayesian (default) and Random tuning strtegies

Now lets try HPO with random strategy.  In the default tuning method above, uses Bayesian Optimization which is a sequential algorithm that learns from past trainings as the tuning job progresses. This highly limits the level of parallelism.

However, with random search you have an option to run all jobs parallel. But a disadvantage is that the random search typically requires running considerably more training jobs to reach a comparable model quality.

Please take into consideration your AWS account limit for the number of instances you can launch in parallel.

In [12]:
#Define the Tuner object
random_tuner = HyperparameterTuner(
    estimator_hpo, 
    objective_metric_name, 
    hyperparameter_ranges, 
    max_jobs=20,  #Total training jobs to be run as part of HPT
    max_parallel_jobs=10, #Maximum number of training jobs to run in parallel
    strategy="Random",
    objective_type = objective_type
)

In [13]:
#Fit method
random_tuner.fit({'train': train_input, 'validation': validation_input})

..............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................!


In [14]:
tuning_job_name = random_tuner.latest_tuning_job.name
tuning_job_name

'sagemaker-xgboost-210808-0153'

In [15]:
##Get results from HPT
tuner_results = sagemaker.HyperparameterTuningJobAnalytics(tuning_job_name)
##Get the results as a dataframe
full_df = tuner_results.dataframe()
##Display results sorted by objective metric values
if len(full_df) > 0:
    df = full_df[full_df["FinalObjectiveValue"] > -float("inf")]
    if len(df) > 0:
        df = df.sort_values("FinalObjectiveValue", ascending=True)
        print("Number of training jobs with valid objective: %d" % len(df))
        print({"lowest": min(df["FinalObjectiveValue"]), "highest": max(df["FinalObjectiveValue"])})
        pd.set_option("display.max_colwidth", -1)  # Don't truncate TrainingJobName
    else:
        print("No training jobs have reported valid results yet.")

df

Number of training jobs with valid objective: 19
{'lowest': 0.4267500042915344, 'highest': 0.5597400069236755}


/home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages/ipykernel/__main__.py:12: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.


alpha       eta  max_depth  min_child_weight  \
1   1.902829  0.601191  5.0        4.747202           
17  0.719768  0.761361  6.0        8.864891           
15  0.596016  0.675929  5.0        9.554632           
19  0.744367  0.893158  7.0        4.708553           
20  1.460550  0.251139  5.0        3.390766           
9   0.445725  0.656241  6.0        9.640180           
6   0.218374  0.237637  5.0        6.574494           
0   0.422439  0.180482  7.0        4.885695           
8   0.980630  0.997615  8.0        8.409500           
14  0.141610  0.747627  3.0        9.640702           
5   0.146491  0.398218  3.0        2.297153           
10  0.535305  0.988762  2.0        4.195519           
4   0.542603  0.734742  10.0       9.672570           
7   0.063834  0.693111  10.0       9.166035           
2   1.072700  0.515337  9.0        3.538788           
12  0.504268  0.234761  2.0        7.609624           
13  1.417817  0.206997  3.0        3.923939           
16  1.445562  0.201184  3.0        8.677592           
3   1.435602  0.099454  6.0        8.237198           

                               TrainingJobName TrainingJobStatus  \
1   sagemaker-xgboost-210808-0153-020-57439525  Completed          
17  sagemaker-xgboost-210808-0153-004-d496afd0  Completed          
15  sagemaker-xgboost-210808-0153-006-9ba28712  Completed          
19  sagemaker-xgboost-210808-0153-002-d76adc9e  Completed          
20  sagemaker-xgboost-210808-0153-001-045796a8  Completed          
9   sagemaker-xgboost-210808-0153-012-a34cca96  Completed          
6   sagemaker-xgboost-210808-0153-015-e6e9bbc6  Completed          
0   sagemaker-xgboost-210808-0153-021-6229bcab  Completed          
8   sagemaker-xgboost-210808-0153-013-4de2cdae  Completed          
14  sagemaker-xgboost-210808-0153-007-c1a4821d  Completed          
5   sagemaker-xgboost-210808-0153-016-980238ac  Completed          
10  sagemaker-xgboost-210808-0153-011-ae8f8ed8  Completed          
4   sagemaker-xgboost-210808-0153-017-14377379  Completed          
7   sagemaker-xgboost-210808-0153-014-93538799  Completed          
2   sagemaker-xgboost-210808-0153-019-9ec530ab  Completed          
12  sagemaker-xgboost-210808-0153-009-2e04645f  Completed          
13  sagemaker-xgboost-210808-0153-008-f7b0b8af  Completed          
16  sagemaker-xgboost-210808-0153-005-26a49a19  Completed          
3   sagemaker-xgboost-210808-0153-018-74fe5580  Completed          

    FinalObjectiveValue         TrainingStartTime           TrainingEndTime  \
1   0.42675             2021-08-08 02:22:55+00:00 2021-08-08 02:42:21+00:00   
17  0.43923             2021-08-08 01:56:16+00:00 2021-08-08 02:17:05+00:00   
15  0.46115             2021-08-08 01:56:23+00:00 2021-08-08 02:15:36+00:00   
19  0.46333             2021-08-08 01:56:39+00:00 2021-08-08 02:18:23+00:00   
20  0.47879             2021-08-08 01:56:16+00:00 2021-08-08 02:15:46+00:00   
9   0.47978             2021-08-08 02:15:41+00:00 2021-08-08 02:36:08+00:00   
6   0.47994             2021-08-08 02:17:25+00:00 2021-08-08 02:37:04+00:00   
0   0.49743             2021-08-08 02:49:15+00:00 2021-08-08 03:10:42+00:00   
8   0.49955             2021-08-08 02:17:10+00:00 2021-08-08 02:40:22+00:00   
14  0.53065             2021-08-08 01:56:21+00:00 2021-08-08 02:15:07+00:00   
5   0.53174             2021-08-08 02:18:15+00:00 2021-08-08 02:36:44+00:00   
10  0.53189             2021-08-08 02:03:26+00:00 2021-08-08 02:20:26+00:00   
4   0.54312             2021-08-08 02:18:27+00:00 2021-08-08 02:43:13+00:00   
7   0.54495             2021-08-08 02:17:13+00:00 2021-08-08 02:41:46+00:00   
2   0.54639             2021-08-08 02:21:03+00:00 2021-08-08 02:44:27+00:00   
12  0.54718             2021-08-08 01:56:30+00:00 2021-08-08 02:13:18+00:00   
13  0.55262             2021-08-08 01:56:24+00:00 2021-08-08 02:14:39+00:00   
16  0.55483             2021-08-08 01:56:27+00:00 2021-08-08 02:14:53+00:00   
3   0.55974             2021-08-08 02:19:34+

The tuning job with random strategy took roughly 60 minutes. Best validation rmse observed is 0.42 with max_depth = 7, alpha = 1.13, eta = 0.22, min_child_weight = 1.02.  

Comparing the times for the tuning with bayesian and random strategy, you see that the random strategy finished about  60 minutes early.  We can see a similar objective value (~0.42), but it took 20 differnet training jobs as opposed to just 10 training jobs with the bayesian strategy.